src/vmm/src/device_manager/legacy.rs

```diff
@@ -85,27 +89,51 @@ impl PortIODeviceManager {
             .map_err(Error::BusError)?;
         self.io_bus
             .insert(
-                Arc::new(Mutex::new(devices::legacy::Serial::new_sink(
-                    self.com_evt_2_4.try_clone().map_err(Error::EventFd)?,
-                ))),
+                Arc::new(Mutex::new(SerialDevice {
+                    serial: Serial::with_events(
+                        self.com_evt_2_4.try_clone().map_err(Error::EventFd)?,
+                        SerialEventsWrapper {
+                            metrics: METRICS.uart.clone(),
+                            buffer_ready_event_fd: None,
+                        },
+                        Box::new(std::io::sink()),
+                    ),
+                    input: None,
+                })),
```

Comment:
nit: maybe you can use some temporary variables to simplify this code block. You might even want to dedupe it as it is repeated a few times.